### ISSUE TITLE
How TorchSharp & Onnx can address the pain points of the ~900 ML.NET Apr2021 survey responses?

### ISSUE THREAD

**Issue by GeorgeS2019** [dotnet/machinelearning#5874]
**Is your feature request related to a problem? Please describe.**
===============================================
The pain points of [the Apr 2021 ML.NET survey](https://devblogs.microsoft.com/dotnet/ml-net-survey-machine-learning-in-net/) and the [result discussions](https://devblogs.microsoft.com/dotnet/ml-net-june-updates-model-builder/)

![image](https://user-images.githubusercontent.com/49812372/125118311-d3ca7b00-e0ef-11eb-99fe-3307f73b936f.png)

**Describe the solution you'd like**
===============================================
It is clear that **NLP** is high on priority
This means **more deep learning NLP use cases ** e.g. using **ML.NET** to load pretrained Hugging Face transformer models using **OnnxRuntime**

We can do that by [porting some of the PyTorch NLP Transformer codes to c#](https://github.com/microsoft/onnxruntime/issues/8345) to address the pain points of ML.NET!!
- [Inference_GPT2_with_OnnxRuntime_on_CPU.ipynb](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/notebooks/Inference_GPT2-OneStepSearch_OnnxRuntime_CPU.ipynb)
- [Inference_GPT2-OneStepSearch_OnnxRuntime_CPU.ipynb](Inference_GPT2-OneStepSearch_OnnxRuntime_CPU.ipynb)

The porting process is now more feasible due to [the recent renaming effort of TorchSharp] (https://github.com/xamarin/TorchSharp/issues/308#issuecomment-877349861) which make **TorchSharp codes MORE resemble PyTorch**

**Comment by michaelgsharp** [comment]
@briacht can you take a look at this?

**Comment by briacht** [comment]
Yes! I had a conversation with @GeorgeS2019 about deep learning in .NET.

As part of our deep learning plan, we will enable NLP scenarios.

I believe these suggestions will start at the TorchSharp level.

**Comment by GeorgeS2019** [comment]
@michaelgsharp @briacht 
Just identified an [error ](https://github.com/dotnet/machinelearning/issues/5894)in the c# example for the method 2 of [OnnxCatalog.ApplyOnnxMode](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.onnxcatalog.applyonnxmodel?view=ml-dotnet#Microsoft_ML_OnnxCatalog_ApplyOnnxModel_Microsoft_ML_TransformsCatalog_System_String_System_Collections_Generic_IDictionary_System_String_System_Int32____System_Nullable_System_Int32__System_Boolean_)l

The  method 2 involves shapeDictionary, which is particularly useful for working with **variable** dimension inputs and outputs.

All examples provided by OnnxCatalog.ApplyOnnxMode use [Image Classification through squeezenet onnx from Onnx zoo models](https://github.com/onnx/models/tree/master/squeezenet) 

The users of the April survey requested NLP use case. 

There are more image related examples involving ML.NET than NLP.  Image use case, unlike NLP, which often does not involve variable dimension inputs and outputs. 

I suggest the [document](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.onnxcatalog.applyonnxmodel?view=ml-dotnet#Microsoft_ML_OnnxCatalog_ApplyOnnxModel_Microsoft_ML_TransformsCatalog_System_String_System_Collections_Generic_IDictionary_System_String_System_Int32____System_Nullable_System_Int32__System_Boolean_) provides, in addition to image, NLP examples from e.g. Onnx Zoo model e.g. [GPT-2](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2) => which will involve variable dimension and the need to sue ShapeDictionary.

**Comment by briacht** [comment]
Thanks for the suggestion @GeorgeS2019!

@luisquintanilla can we add an issue to the docs repo for this?

**Comment by GeorgeS2019** [comment]
@briacht 

As correctly [pointed out](https://github.com/dotnet/machinelearning/issues/5273#issuecomment-652638075) by @antoniovs1029
> ML.NET **doesn't** currently have any transformer to do **tensor reshaping**, and it's necessary for users to actually implement their own reshape logic

This missing feature has been raised by @yaeldekel  here [Add "Reshape Transform"](https://github.com/dotnet/machinelearning/issues/765)

=> I request to look into this "tensor reshaping" and the **variable dimension** discussed above. Perhaps both are related.

Here are **NLP cases** applying [OnnxCatalog.ApplyOnnxModel](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.onnxcatalog.applyonnxmodel?view=ml-dotnet#Microsoft_ML_OnnxCatalog_ApplyOnnxModel_Microsoft_ML_TransformsCatalog_System_String_System_Collections_Generic_IDictionary_System_String_System_Int32____System_Nullable_System_Int32__System_Boolean_) onnx models

- [Error on prediction with LSTM Model in ONNX Format in ML.Net](https://github.com/dotnet/machinelearning/issues/5273)
- [LSTM Data Input](https://github.com/dotnet/machinelearning-samples/issues/828)

Perhaps by implementing the "Reshape Transform", this could address challenges when working with NLP with ML.NET?

**Comment by matouskozak** [comment]
Triage: Closing as outdated.
